### ISSUE TITLE
Toward more props types

### ISSUE THREAD

**Issue by mbovel** [hashml/hashml#34]
TLDR; I think we are on the right track with our IR. The schema should not handle much more of the transformations the application needs (like indexing). The only thing our IR misses is more precise types. In the schema, these types should be defined in child tags and not in the parents.

# 1. Current situation

Before going further, here is a tentative to summarize the current architecture of our props schemas.

## 1.1. Block props

Right now, we have two block props types:

- Parsed props, which group children with specific tags in an array.
- Raw props, meaning that the whole body of the parent block should not be parsed as hashml, but returned as an array of lines.

### 1.1.1. Parsed block props

Parsed block props are defined in accordance to the following interface:

```typescript
export interface BlockPropDefinition {
	name: string;
	raw?: undefined;
	content: BlockPropContentDefinition[];
}
```

and might generate a key/pair like in an object like:

```typescript
"children": [{tag: "section", …}, {tag: "paragraph", …}]
```

**Note**: a definition of “prop” might be: “something that consume some child tags and generates a key/value pair”.

### 1.1.2. Raw block props

 Raw block props have the following definition:

```typescript
export interface RawBlockPropDefinition {
	name: string;
	raw: true;
}
```

and might generate a key/pair like:

```typescript
"content": ["console.log(\"Hashml\")", "console.log(\"rocks!\")"]
```

**Note:** only one raw prop might be defined for a given block, and if it is, it must be the only prop (as it means that the whole body should not be parsed).

## 1.2. Inline props

Similarly to block props, there are two inline props types:

- Parsed props, which means that the given inline span should be parsed as hashml.
- Raw props, meaning that the whole inline span should just be returned as a string.

**Note:** here I use the term “inline span” to refer to the part of a line which is either the head of a block, or an argument. Once parser, this is an “inline group”.

### 1.2.1. Parsed inline props

Parsed inline props have the following interface:

```typescript
export interface ParsedInlinePropDefinition {
	name: string;
	raw?: undefined;
	content: string[];
}
```

And generate an array of string and child inline tags (this is an “inline group”), like for example:

```typescript
["Lorem", {tag: "strong", args: {content: ["Ipsum"]}}, "dolor"]
```

### 1.2.2. Raw inline props

Raw inline props have the following interface:

```typescript
export interface RawInlinePropDefinition {
	name: string;
	raw: true;
}
```

Which for now generate arrays of a single string like:

```
["#notATag[this is raw"]
```

## 1.3. IR representation

In the intermediate representation, block and inline tags are represented alike with the following interface:

```typescript
export interface IRNode {
	tag: string;
	props: {
		[name: string]: IRNodeList;
	};
}
```

# 2. Potential problems with the current situation

1. Most applications will need to **transform the data further**, for example to parse a given string as a number or as a date. Doing it at the application level yields two problems:
	1. **Code duplication**: each application might have to reinvent the wheel. Typically, we wouldn't want the code to parse a number or a date to be rewritten by each application.
	2. **Error handling**: handling parser errors such as string to date transformation in the application makes the error handling more complex.
2. Having more precise data types would avoid creating tons of useless arrays and objects. For example instead of having `{..., id: [{tag: id, props: {content: ["my_id"]}}]}`, we could just have `{..., id: "my_id"}` if it was specified that the `id` tag must be interpreted as a string prop. This is both great for performance and humane usage.

# 3. Goals and non-goals

I think the central questions are “What representations do we actually need?”., and “Where should they be generated?”.

## 3.1 What representations do we actually need?

At the beginning of the project, we only had an AST, now we have an IR, and by working on Hashdocs, it's obvious that the IR is still not enough. We need more than just _data_: for example we need indexes for IDs and refs, etc.

So, at this stage I identified six different representations useful for Hashdocs:

1. Text: the input. Obviously this need to be processed further.

	```typescript
	"#section My section\nhello"
	```
2. The AST, which is a direct tree representation of the text, where nodes are still loosely typed.

	```typescript
	{
	tag: "section",
	head: ["My section"],
	children: {
		tag: undefined,
		head: ["hello"]
	}
	}
	```
3. The IR, a more machine-friendly, data-oriented representation, where each node have/should have a precise type. This representation is useful to be storage and real-time communication (OT, CRDTs).

	Right now:

	```typescript
	{
		tag: "section",
		props: {
		title: ["My section"],
		id: [],
		content: [
			{
			tag: "paragraph",
			props: { content:  ["hello"] }
			}
		]
		}
	}
	```

	With more precise types, it could be for example:

	```typescript
	{
		tag: "section",
		title: ["My section"],
		id: undefined,
		content: [
		{
			tag: "paragraph",
			content:  ["hello"]
		}
		]
	}
    ```
4. The ““Document Object Model”” (or whatever we want to call it): the IR augmented with indexes, computed values, etc. For example:

	```typescript
	{
		tag: "section",
		title: ["My section"],
		titleString: "My section", // string representation of title for references, TOC, etc.
		id: undefined,
		idObject: ID("my_section_2"), // id object generated from id tag if present, or from title otherwise
		subsection: [...], // section children, for TOC,
		mathContext: {...}, // maths definitions, etc.
		content: [
		{
			tag: "paragraph",
			content:  ["hello"]
		}
		]
	}
	```
5. React components.
6. Final HTML text or DOM output.

## 3.2 Where should they be generated

We noticed that we could skip the AST and generate directly the IR from the input text. I think this makes sense.

For a long time, I thought that we could directly jump the the next representation as well: output something that is “ready to be used” by the application, like the DOM (point 4) above. But more I think about it, more it seems that it is too much for our schema system: we just cannot anticipate all potential indexing and other custom transformations any application might need.

All of this to say, I think our current IR is pretty much the right representation output to expect from our schema system. As discussed already, the only I would suggest is the ability to produce more precise type (like in second part of point 3. above: “With more precise types, it could be […]”).

# 4. Design

## 4.1 New IRNode interface 

The IR interface should be changed to:

```typescript
export interface IRNode {
	tag: string;
	[name: string]: Array<string|IRNode> | string | IRNode | undefined;
}
```

This is the most general possible type for IRNode, for internal use, but more precise types can be generated from the schema for further use, as you already started to implement :) For example:

```typescript
	export interface BlockSection {
	tag: string;
	content: Array<BlockSection|BlockParagraph>;
	id: string?;
	authors: Array<BlockAuthor>;
	title: Array<string|InlineStrong|InlineEmphasis>;
}
```

## 4.2 Schema interface

I am still not sure exactly how the schema for such type transformations should look like,
 but I have an idea like:

````typescript
const schema = {
	"blocks": {
		"section": {
			type: "object",
			props: {
				// VERY IMPORTANT: no type information here, I think the types of “title”, “id” or “subtitle”
				// should be defined in child definitions, not in the parent. This allows for better modularity:
				// if the type of "title" change, we don't need to change it in every block it appears.
				title: head(optional("title")), // -> Array<string|Inline>?
				subtitle: optional("subtitle"), // -> Array<string|Inline>?
				id: optional("id"), // -> string?
				content: [zeroOrMore("paragraph"), zeroOrMore("section")], // -> Array<BlockParagraph|BlockSection>
				authors: zeroOrMore("author") // -> Array<BlockAuthor>
			}
		},
		"title": {
			type: "inline" // kind of like “line tag”: only the head is interpreted, as inline content
		},
		"subtitle": {
			type: "inline",
		},
		"author": {
			type: "object",
			props: {
				name: head(optional("name")),
				website: one("website"),
				birthday: one("birthday")
			}
		},
		"name": {
			type: "inline"
		},
		"id": {
			type: "string" // only the head is interpreted, as string (raw)
		},
		"website": {
			type: "url" // only the head is interpreted, as url
		},
		"birthday": {
			type: "date"
		}
	},
	"inline": {
		// I am pretty happy for how the schema look for blocks. For inline definitions, I am less sure.
		"strong": {
			"args": [
				["content", {type: "inline"}]
			]
		},
		"link": {
			"args": [
				["url", {type: "url"}],
				["url", {type: "inline"}]
			]
		}
	}
}
````

I think this could be implemented incrementally by modifying our current system step by step.

**Comment by MaximeKjaer** [comment]
I've thought about all of this today, and have some feedback on your suggestions, as well as suggestions of my own to move forward.

Feedback:

- **`type` field**: 
  - It's a little verbose, but the verbosity can be eliminated by a good set of factory functions. 
  - Having this field also allows us to make discriminated union types, which is a nicer solution than the current `raw?: undefined` :wink: 
  - Overall, I like this :+1: It adds clarity and will make Typescript generation easier.
- **`type: "inline"`**:
  - I don't think that this is the best name, because it clashes with inline blocks (e.g. `#bold[text]`).
  - I think we should call it `line`. This matches the terminology we've had for a long time about "line tags", "block tags", etc.
- **Types in child definition**: I don't agree with this.
  - When describing *actual* children of a block, then yes, the children define the type.
  - When describing the head, I don't think it makes sense. In section 4.2, for instance, it forces us to introduce things that aren't *really* block tags, but rather just an artifact of wanting to have a named head (e.g. the "title" tag). And this indirection is a little weird, because all the other tags define the type of their head locally. So why should "object" blocks do that?
  - I don't think that the use case you're describing (where you want to change the type of all heads called `title`, say from `string` to `inline`) is that common. In a situation where other people depend on your schema, this can be a breaking change too, so I think it's good that the big, rare change requires more work.
  - The more common use case, I think, is to add a new inline tag for example, and allow it in a bunch of tag heads. For example, say you want to introduce a `#blink` tag: you must now allow it in section titles, image captions, etc. In this case, the mechanism we need is not to defer to a virtual tag (which I think is a hack), but instead a "macro system" to keep things DRY. In TS, this problem is solved by having a `const group = [...]`, but it would need syntax in a schema written in Hashml.

Suggestion for going forward:

- All tags define the type of their own head or args, without indirection.
- `type: "object"` block tags list their props, which list children and cardinalities.
- Inline tags list their arguments, which list allowed content.

I've started an implementation on the `schema-types` branch.
